File: advent_of_python/src/solvers_2015/solver_03.py
from typing import Any

from solver import Solver


class SolverImpl(Solver):

    def solver_part_1(self, data: str) -> Any:
        location = (0, 0)
        visited: set[tuple[int, int]] = {location}
        for character in data:
            location = self._apply_direction(character, location)
            visited.add(location)
        return len(visited)

    def _apply_direction(self, character, location):
        match character:
            case "^":
                location = (location[0], location[1] + 1)
            case "v":
                location = (location[0], location[1] - 1)
            case ">":
                location = (location[0] + 1, location[1])
            case "<":
                location = (location[0] - 1, location[1])
        return location

    def solve_part_2(self, data: str) -> Any:
        locations = [(0, 0), (0, 0)]
        index = 0
        visited: set[tuple[int, int]] = {locations[0]}
        for character in data:
            locations[index] = self._apply_direction(character, locations[index])
            visited.add(locations[index])
            index = 1 - index
        return len(visited)

File: advent_of_go/src/solvers_2015/solver_03.go
package solvers_2015

import "github.com/hashicorp/go-set/v3"

type Solver03 struct{}

type Solver03Location struct {
	X int
	Y int
}

func (Solver03) SolvePart1(data string) int {
	location := Solver03Location{0, 0}
	visited := set.From[Solver03Location]([]Solver03Location{location})
	for _, c := range data {
		switch c {
		case '^':
			location.Y++
		case 'v':
			location.Y--
		case '>':
			location.X++
		case '<':
			location.X--
		}
		visited.Insert(location)
	}
	return visited.Size()
}

func (Solver03) SolvePart2(data string) int {
	locations := []Solver03Location{{0, 0}, {0, 0}}
	visited := set.From[Solver03Location]([]Solver03Location{locations[0]})
	index := 0
	for _, c := range data {
		location := locations[index]
		switch c {
		case '^':
			location.Y++
		case 'v':
			location.Y--
		case '>':
			location.X++
		case '<':
			location.X--
		}
		visited.Insert(location)
		locations[index] = location
		index = 1 - index
	}
	return visited.Size()
}
